lib/bamboo/test.ex

```diff
@@ -90,23 +94,91 @@ defmodule Bamboo.Test do
       assert_delivered_email(unsent_email) # Will fail
   """
   def assert_delivered_email(%Bamboo.Email{} = email) do
-    import ExUnit.Assertions
     email = Bamboo.Mailer.normalize_addresses(email)
-    assert_received {:delivered_email, ^email}
+
+    do_assert_delivered_email(email)
+  end
+
+  defp do_assert_delivered_email(email) do
+    receive do
+      {:delivered_email, ^email} -> true
+    after
+      @timeout -> flunk_with_email_list(email)
+    end
+  end
+
+  defp flunk_with_email_list(email) do
+    if Enum.empty?(delivered_emails) do
+      flunk """
+      There were 0 emails delivered to this process.
+
+      If you expected an email to be sent, try these ideas:
+
+        1) Make sure you call deliver_now/1 or deliver_later/1 to deliver the email
+        2) Make sure you are using the Bamboo.TestAdapter
+        3) Use the process name feature of Bamboo.Test. This will allow Bamboo.Test
+           to work across processes: use Bamboo.Test, process_name: :my_test_name
+        4) If you are writing an acceptance test through a headless browser, see
+           if writing a controller test would be possible instead. Otherwise try
+           option 3.
+      """
+    else
+      flunk """
+      There were no matching emails.
+
+      No emails matched:
+
+        #{inspect email}
+
+      Delivered emails:
+
+      #{delivered_emails_as_list}
```

Comment:
The problem was that if I indented in this string it would indent the first email, but the rest wouldn’t be indented. So I had to do the indenting in the delivered_emails_as_list function. Open to other ideas if you can think of something.

> On Apr 8, 2016, at 11:13 AM, Josh Steiner <notifications@github.com> wrote:
> 
> In lib/bamboo/test.ex <https://github.com/thoughtbot/bamboo/pull/109#discussion_r59040183>:
> 
> > +           to work across processes: use Bamboo.Test, process_name: :my_test_name
> > +        4) If you are writing an acceptance test through a headless browser, see
> > +           if writing a controller test would be possible instead. Otherwise try
> > +           option 3.
> > +      """
> > +    else
> > +      flunk """
> > +      There were no matching emails.
> > +
> > +      No emails matched:
> > +
> > +        #{inspect email}
> > +
> > +      Delivered emails:
> > +
> > +      #{delivered_emails_as_list}
> Should this be indented too?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/thoughtbot/bamboo/pull/109/files/6b0e3abf62fa061c8fd7187bf78494cd729edb8a#r59040183>